src/Hangman/Puzzle.hs

```diff
@@ -32,28 +32,30 @@ stateAnnouncement puzzle =
         _ -> ""
 
 handleInput :: Puzzle -> String -> Puzzle
+handleInput puzzle [c] =
+    handleGuess puzzle c
 handleInput puzzle input =
-    case input of
-        [c] -> handleGuess puzzle c
-        _   -> puzzle { state = InvalidGuess input }
+    puzzle { state = InvalidGuess input }
 
 handleGuess :: Puzzle -> Char -> Puzzle
-handleGuess puzzle guess =
+handleGuess puzzle@(Puzzle _ completed' guesses' _) guess =
     newPuzzle { state = newState }
   where
-    (Puzzle _ completed' guesses' _) = puzzle
-    newPuzzle = puzzle { guesses = (guess : guesses'), completed = newFilledInSoFar }
+    newPuzzle = puzzle { guesses = (guess : guesses'), completed = newFilledInSoFar puzzle guess }
     newState
         | all isJust (completed newPuzzle) = Won
         | length (guesses newPuzzle) > 7   = Lost
         | elem guess guesses'              = AlreadyGuessed guess
         | otherwise                        = Playing
+
+newFilledInSoFar :: Puzzle -> Char -> [Maybe Char]
+newFilledInSoFar (Puzzle answer' completed' _ _) guess =
```

Comment:
By the way, since the field names aren't used as functions in `newFilledInSoFar`, these could be `answer` and `completed`, even though that "conflicts" with the record functions.